src/clang-c-frontend/clang_c_convert.cpp

```diff
@@ -1444,7 +1447,50 @@ bool clang_c_convertert::get_expr(const clang::Stmt &stmt, exprt &new_expr)
     if(get_expr(*compound.getInitializer(), initializer))
       return true;
 
-    new_expr = initializer;
+    typet t = initializer.type();
+
+    /* A compound literals is an LValue that has associated static or automatic
+     * storage, depending on whether it appears in file or block scope.
+     * Therefore, in C, for instance a pointer to it can be taken and used
+     * within the same block without restrictions, e.g.
+     *
+     *   int *p = &(int){0}, x = 42;
+     *   memcpy(p, &x, sizeof(int));
+     *   assert(*p == 42);
+     *
+     * It has the same semantics as an non-anonymous variable of the
+     * corresponding type. Thus, we introduce a new symbol to represent it.
+     */
+
+    /* "cl" stands for "compound literal", but that is too long for possibly
+     * frequent copies of the symbol's name and id. */
+    symbolt &cl = anon_symbol.new_symbol(context, initializer.type(), "cl$");
+    /* mimic get_default_symbol() w/o setting id and name as that's already
+     * taken care of by anon_symbol.new_symbol() above */
+    cl.mode = "C";
+    cl.module = get_modulename_from_path(location.file().as_string());
+    cl.location = location;
+    cl.type = t;
+
+    cl.static_lifetime = compound.isFileScope();
+    cl.is_extern = false;
+    cl.file_local = true;
+    cl.value = initializer;
+
+    new_expr = symbol_expr(cl);
+
+    if(current_block)
```

Comment:
I'm not sure about this piece of code here, maybe this can be converted to a codet, using `convert_expression_to_code` instead?

Reply:
This piece of code is inserting a declaration of the new symbol into the current block before the statement the currently converted expression belongs to. It basically follows `get_var()` here: https://github.com/esbmc/esbmc/blob/038c09a042e6d86d5682bee9cf8f20c1300bb9ef/src/clang-c-frontend/clang_c_convert.cpp#L481-L493 As discussed in https://github.com/esbmc/esbmc/pull/644#discussion_r827917073, local symbols (or those with automatic storage) are required to be declared in order for the value-set analysis to recognize them as being local to a function.